controllers/cloudstackmachine_controller.go

```diff
@@ -70,93 +74,95 @@ func (r *CloudStackMachineReconciler) Reconcile(ctx context.Context, req ctrl.Re
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
 	// Setup patcher. This ensures modifications to the csMachine copy fetched above are patched into the origin.
-	if patchHelper, retErr := patch.NewHelper(csMachine, r.Client); retErr != nil {
-		return ctrl.Result{}, retErr
+	if patchHelper, err := patch.NewHelper(csMachine, r.Client); err != nil {
+		return ctrl.Result{}, err
 	} else {
 		defer func() { // If there was no error on return, but the patch fails, set the error accordingly.
-			if err := patchHelper.Patch(ctx, csMachine); retErr == nil && err != nil {
-				retErr = err
+			if err = patchHelper.Patch(ctx, csMachine); err != nil {
+				msg := "error patching CloudStackMachine %s/%s"
+				err = errors.Wrapf(err, msg, csMachine.Namespace, csMachine.Name)
+				retErr = multierror.Append(retErr, err)
 			}
 		}()
 	}
 
 	// Fetch the CAPI Machine.
-	machine, retErr := util.GetOwnerMachine(ctx, r.Client, csMachine.ObjectMeta)
-	if retErr != nil {
-		return ctrl.Result{}, retErr
-	} else if machine == nil {
+	capiMachine, err := util.GetOwnerMachine(ctx, r.Client, csMachine.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	} else if capiMachine == nil {
 		log.Info("Waiting for CAPI cluster controller to set owner reference on CloudStack machine.")
 		return ctrl.Result{RequeueAfter: RequeueTimeout}, nil
 	}
 
 	// Fetch the CAPI Cluster.
-	cluster, retErr := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
-	if retErr != nil {
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, capiMachine.ObjectMeta)
+	if err != nil {
 		log.Info("Machine is missing cluster label or cluster does not exist.")
 		return ctrl.Result{RequeueAfter: RequeueTimeout}, nil
 	}
 
-	// Check the machine is not paused.
+	// Check if the machine is paused.
 	if annotations.IsPaused(cluster, csMachine) {
 		log.Info("CloudStackMachine or linked Cluster is paused. Requeuing reconcile.")
 		return reconcile.Result{}, nil
 	}
 
-	// Delete VM instance if deletion timestamp present.
-	if !csMachine.DeletionTimestamp.IsZero() {
-		return r.reconcileDelete(log, csMachine)
-	}
-
 	// Fetch the CloudStack cluster associated with this machine.
 	csCluster := &infrav1.CloudStackCluster{}
-	if retErr := r.Client.Get(
+	if err := r.Client.Get(
 		ctx,
 		client.ObjectKey{
 			Namespace: csMachine.Namespace,
 			Name:      cluster.Spec.InfrastructureRef.Name},
-		csCluster); retErr != nil {
-		if client.IgnoreNotFound(retErr) == nil {
+		csCluster); err != nil {
+		if client.IgnoreNotFound(err) == nil {
 			log.Info("CloudStackCluster not found.")
 			return ctrl.Result{RequeueAfter: RequeueTimeout}, nil
 		} else {
-			return ctrl.Result{}, retErr
+			return ctrl.Result{}, err
 		}
 	} else if csCluster.Status.ZoneID == "" {
 		log.Info("CloudStackCluster ZoneId not initialized. Likely not ready.")
 		return ctrl.Result{RequeueAfter: RequeueTimeout}, nil
 	}
 
+	// Delete VM instance if deletion timestamp present.
```

Comment:
This "delete VM" code block was moved above the "Fetch Cloudstack Cluster" block in https://github.com/aws/cluster-api-provider-cloudstack/commit/2794cc800aad7fa206b21444d6bd6802c8f7408d to address a bug where machine deletion fails if cluster has already been deleted.

Reply:
Overall, I believe this to be a false concern. The cloudstack cluster has owner refs all the way down to the machine, and can't be deleted until the machines say they're deleted.

Still, I'm fine moving it back.

I originally needed `csCluster` to specify Account in deletion and fetch calls, but in retrospect I don't think that was necessary, so I moved this back where it was.